internal/corazawaf/rule.go

```diff
@@ -282,12 +288,19 @@ func (r *Rule) doEvaluate(phase types.RulePhase, tx *Transaction, cache map[tran
 			matchedValues = append(matchedValues, matchedChainValues...)
 			nr = nr.Chain
 		}
+		// Flow actions are evaluated also if the rule engine is set to DetectionOnly
+		for _, a := range r.actions {
+			if a.Function.Type() == rules.ActionTypeFlow {
+				tx.WAF.Logger.Debug("[%s] [%d] Evaluating flow action %s for rule %d", tx.id, rid, a.Name, r.ID_)
+				a.Function.Evaluate(r, tx)
+			}
+		}
 		// we need to add disruptive actions in the end, otherwise they would be triggered without their chains.
```

Comment:
Just wondering is it important for flow actions to be evaluated before disruptive actions? Or perhaps they are actually supposed to be evaluated in the order of the rule? If it's ok to do it, I would just loop once and check the rule engine type inside the loop instead of outside

Reply:
They should be evaluated in the order they are parsed. We had a similar discussion with @fzipi about giving priority to `allow` action in the end it is the order being parsed. See https://github.com/corazawaf/coraza/pull/527/files#diff-1b7cbdff9e4ff0a0bb60009d3864bc5d7e7fae7a351aa525176466834e8a70e0R302

Reply:
Thank you, moved to just one loop and tweaked the parser so that only one disrupting action per rule will be executed. This enforces the following statement (from both Coraza and ModSec doc):
```
There can only be one disruptive action per rule (if there are multiple disruptive actions present, or inherited, only the last one will take effect), or rule chain (in a chain, a disruptive action can only appear in the first rule).
```